daemon/cmd/policy.go

```diff
@@ -58,20 +63,90 @@ func (d *Daemon) initPolicy(
 	}
 	d.datapathRegenTrigger = rt
 
-	d.policy = policy.NewPolicyRepository(d.identityAllocator,
-		d.identityAllocator.GetIdentityCache(),
-		certManager,
-		secretManager,
+	d.monitorAgent.RegisterNewConsumer(authManager)
+	return nil
+}
+
+type policyParams struct {
+	cell.In
+
+	Lifecycle       hive.Lifecycle
+	EndpointManager endpointmanager.EndpointManager
+	CertManager     certificatemanager.CertificateManager
+	SecretManager   certificatemanager.SecretManager
+	Datapath        datapath.Datapath
+	CacheStatus     k8s.CacheStatus
+}
+
+type policyOut struct {
+	cell.Out
+
+	IdentityAllocator CachingIdentityAllocator
+	Repository        *policy.Repository
+	Updater           *policy.Updater
+	IPCache           *ipcache.IPCache
+}
+
+// newPolicyTrifecta instantiates CachingIdentityAllocator, Repository and IPCache.
+//
+// The three have a circular dependency on each other and therefore require
+// special care.
+func newPolicyTrifecta(params policyParams) (policyOut, error) {
+	iao := &identityAllocatorOwner{}
+	idAlloc := &cachingIdentityAllocator{
+		cache.NewCachingIdentityAllocator(iao),
+		nil,
+	}
+
+	iao.policy = policy.NewStoppedPolicyRepository(
+		idAlloc,
+		idAlloc.GetIdentityCache(),
+		params.CertManager,
+		params.SecretManager,
 	)
-	d.policy.SetEnvoyRulesFunc(envoy.GetEnvoyHTTPRules)
-	d.policyUpdater, err = policy.NewUpdater(d.policy, epMgr)
+	iao.policy.SetEnvoyRulesFunc(envoy.GetEnvoyHTTPRules)
+
+	policyUpdater, err := policy.NewUpdater(iao.policy, params.EndpointManager)
 	if err != nil {
-		return fmt.Errorf("failed to create policy update trigger: %w", err)
+		return policyOut{}, fmt.Errorf("failed to create policy update trigger: %w", err)
 	}
+	iao.policyUpdater = policyUpdater
 
-	d.monitorAgent.RegisterNewConsumer(authManager)
+	ctx, cancel := context.WithCancel(context.Background())
 
-	return nil
+	ipc := ipcache.NewIPCache(&ipcache.Configuration{
+		Context:           ctx,
+		IdentityAllocator: idAlloc,
+		PolicyHandler:     iao.policy.GetSelectorCache(),
+		DatapathHandler:   params.EndpointManager,
+		NodeHandler:       params.Datapath.Node(),
+		CacheStatus:       params.CacheStatus,
+	})
+	idAlloc.ipcache = ipc
+
+	params.Lifecycle.Append(hive.Hook{
+		OnStart: func(hc hive.HookContext) error {
+			iao.policy.Start()
+			return nil
+		},
+		OnStop: func(hc hive.HookContext) error {
+			cancel()
```

Comment:
While I don't think it actually matters in correctness, I believe it would be cleaner to call into the shutdown routines first, and then `cancel` the context. For example, some of the shutdown routines (like `ipc.Shutdown`) will wait for a controller to finish it's work first.